### PR TITLE
Added center as top-left for shear X/Y ops for autoaugment

### LINKED ISSUE #5204: There is a Difference between Torchvsion Shear and PIL Shear
### 🐛 Describe the bug

As @agaldran pointed out in the TrivialAugment repo (https://github.com/automl/trivialaugment/issues/6), there seems to be difference between the behavior of a shear implemented with the PIL and the TorchVision affine transform.

This is an issue for the autoaugment algorithms, as this might yield to different results compared to other implementations of these algorithms. It might be an issue for other applications as well. Both the AutoAugment (https://github.com/tensorflow/models/blob/fd34f711f319d8c6fe85110d9df6e1784cc5a6ca/research/autoaugment) and the TrivialAugment (https://github.com/automl/trivialaugment) reference implementations use PIL, while RandAugment has no reference implementation.

```
from PIL import Image
import math
import torchvision # >= 0.11

from torchvision.transforms import functional as F
interpolation = torchvision.transforms.InterpolationMode.NEAREST
fill = None

img = Image.new('RGB', (32,32), (255,255,0))
magnitude = .7

# shear_x as seen in torchvision https://github.com/pytorch/vision/blob/b5aa0915fe16e82ee4c24919032b4e7afae3ae1b/torchvision/transforms/autoaugment.py#L17
im_torch = F.affine(img, angle=0.0, translate=[0, 0], scale=1.0, shear=[math.degrees(magnitude), 0.0],
                    interpolation=interpolation, fill=fill)

# shear_x as seen in https://github.com/automl/trivialaugment/blob/3bfd06552336244b23b357b2c973859500328fbb/aug_lib.py#L156 and https://github.com/tensorflow/models/blob/fd34f711f319d8c6fe85110d9df6e1784cc5a6ca/research/autoaugment/augmentation_transforms.py#L290
im_pil = img.transform(img.size, Image.AFFINE, (1, magnitude, 0, 0, 1, 0))

im_torch.show()
im_pil.show()
```

This script will yield the following images:

download
download

Which should be the same.

It looks like torchvision has a fixed center, while PIL uses a fixed top. I did not dig into the code much, yet, though. Is there someone here who implemented `affine` maybe and can give a reasoning for the different shearing. 

### Versions

Collecting environment information...
PyTorch version: 1.10.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-94-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.0
[pip3] torch==1.10.1
[pip3] torchvision==0.11.2
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.22.0                   pypi_0    pypi
[conda] torch                     1.10.1                   pypi_0    pypi
[conda] torchvision               0.11.2                   pypi_0    pypi

cc @vfdev-5 @datumbox

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #5204

Description:
- Added center as top-left for shear X/Y ops for autoaugment

@datumbox do you think we should add something else here to be fully consistent ?

Here is how affine with center=(0, 0) behaves for shear X  and Y vs using PIL as in TF:
```python
from PIL import Image
import math
import torchvision # >= 0.11
import numpy as np

from torchvision.transforms import functional as F
interpolation = torchvision.transforms.InterpolationMode.NEAREST
fill = None

img = Image.fromarray(np.arange(3 * 128 * 128, dtype='uint8').reshape(128, 128, 3))
magnitude = .4

# shear_x as seen in torchvision https://github.com/pytorch/vision/blob/b5aa0915fe16e82ee4c24919032b4e7afae3ae1b/torchvision/transforms/autoaugment.py#L17
im_torch = F.affine(img, angle=0.0, translate=[0, 0], scale=1.0, shear=[math.degrees(magnitude), 0.0],                    
                    interpolation=interpolation, fill=fill, center=[0, 0])

# shear_x as seen in https://github.com/automl/trivialaugment/blob/3bfd06552336244b23b357b2c973859500328fbb/aug_lib.py#L156 and https://github.com/tensorflow/models/blob/fd34f711f319d8c6fe85110d9df6e1784cc5a6ca/research/autoaugment/augmentation_transforms.py#L290
im_pil = img.transform(img.size, Image.AFFINE, (1, magnitude, 0, 0, 1, 0))

print("Original image")
img.show()
print("Torchvision Shear X with center as top-left")
im_torch.show()
print("Shear X using PIL as TF")
im_pil.show()

# shear_x as seen in torchvision https://github.com/pytorch/vision/blob/b5aa0915fe16e82ee4c24919032b4e7afae3ae1b/torchvision/transforms/autoaugment.py#L17
im_torch = F.affine(img, angle=0.0, translate=[0, 0], scale=1.0, shear=[0.0, math.degrees(magnitude)],                    
                    interpolation=interpolation, fill=fill, center=[0, 0])

# shear_x as seen in https://github.com/automl/trivialaugment/blob/3bfd06552336244b23b357b2c973859500328fbb/aug_lib.py#L156 and https://github.com/tensorflow/models/blob/fd34f711f319d8c6fe85110d9df6e1784cc5a6ca/research/autoaugment/augmentation_transforms.py#L290
im_pil = img.transform(img.size, Image.AFFINE, (1, 0, 0, magnitude, 1, 0))


print("Torchvision Shear Y with center as top-left")
im_torch.show()
print("Shear Y using PIL as TF")
im_pil.show()
```

cc @SamuelGabriel 

EDIT:

Links to TF implementation uses Pillow with shear parametrized such that fixed point is top-left:
- https://github.com/tensorflow/models/blob/885fda091c46c59d6c7bb5c7e760935eacc229da/research/autoaugment/augmentation_transforms.py#L273-L290

However, official vision ops refers as well autoaugment paper and does shear with TF raw ops, the only difference is that they do fill_mode="reflect" by default:
- https://github.com/tensorflow/models/blob/885fda091c46c59d6c7bb5c7e760935eacc229da/official/vision/beta/ops/augment.py#L504-L512
- https://github.com/tensorflow/models/blob/885fda091c46c59d6c7bb5c7e760935eacc229da/official/vision/beta/ops/augment.py#L180-L181
- https://github.com/keras-team/keras/blob/89879e2c76e86c685e44c47a6cdb82f7e645c142/keras/layers/preprocessing/image_preprocessing.py#L654

Output for `magnitude = .4` shear X with TF from official vision ops:
image

### CODE REVIEW

**Review by datumbox — APPROVED** [review]
@vfdev-5 Thanks a lot for the investigations and for sending the PR.

We had long discussions with Victor about this because it's a tough call but I'll summarize here the reasons why we believe we should merge this BC-breaking change. After looking into the original implementations of AutoAugment and RandAugment, it became clear that their Shear method was using the top-left as the center. Unfortunately TorchVision didn't support this option until it was raised by the author of TrivialAugment @SamuelGabriel at #5204. This means that our implementation deviated from the canonical and thus we consider this discrepancy a bug.

Though I don't believe that this discrepancy had major effect on the overall accuracy of the models that we trained previously, I think it's worth aligning TorchVision's behaviour with the rest of the ecosystem as the majority of frameworks and libraries we checked all implement the Shear operator with center the top-left corner.

I recommend leaving this PR unmerged for a couple of days to allow for people to comment. If there are no objections we can merge it early next week.
